smartsim/_core/control/interval.py

```diff
@@ -0,0 +1,115 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import time
+import typing as t
+
+Seconds = t.NewType("Seconds", float)
+
+
+class SynchronousTimeInterval:
+    """A utility class to represent and synchronously block the execution of a
+    thread for an interval of time.
+    """
+
+    def __init__(self, delta: float | None, strict: bool = True) -> None:
+        """Initialize a new `SynchronousTimeInterval` interval
+
+        :param delta: The difference in time the interval represents. If
+            `None`, the interval will represent an infinite amount of time in
+            seconds.
+        :param strict: Wether or not to raise in the case where negative time
+            delta is provided. If `False`, the `SynchronousTimeInterval` and a negative `delta`
+            is provided, the interval will be infinite.
+        :raises ValueError: The `delta` is negative and `strict` is `True`
+        """
+        if delta is not None and delta < 0 and strict:
+            raise ValueError("Timeout value cannot be less than 0")
+        if delta is None or delta < 0:
+            delta = float("inf")
+        self._delta = Seconds(delta)
+        """The amount of time, in seconds the interval spans."""
+        self._start = time.perf_counter()
+        """The time of the creation of the interval"""
+
+    @property
+    def delta(self) -> Seconds:
+        """The difference in time the interval represents
+
+        :returns: The difference in time the interval represents
+        """
+        return self._delta
+
+    @property
+    def elapsed(self) -> Seconds:
+        """The amount of time that has passed since the interval was created
+
+        :returns: The amount of time that has passed since the interval was
+            created
+        """
+        return Seconds(time.perf_counter() - self._start)
+
+    @property
+    def remaining(self) -> Seconds:
+        """The amount of time remaining in the interval
+
+        :returns: The amount of time remaining in the interval
+        """
+        return Seconds(max(self.delta - self.elapsed, 0))
+
+    @property
+    def expired(self) -> bool:
+        """The amount of time remaining in interval
+
+        :returns: The amount of time left in the interval
+        """
+        return self.remaining <= 0
+
+    @property
+    def infinite(self) -> bool:
+        """Is the timeout interval infinitely long.
+
+        :returns: `True` if the delta is infinite, `False` otherwise
+        """
+        return self.remaining == float("inf")
+
+    def new_interval(self) -> SynchronousTimeInterval:
+        """Make a new timeout with the same interval
+
+        :returns: The new timeout
+        """
+        return type(self)(self.delta)
+
+    def wait(self) -> None:
```

Comment:
Do we like the name `wait`? Should it instead be block??

Reply:
Since it performs a block, I think block might be more descriptive of the function?

Reply:
Changed to `block`!